mkosi/qemu.py

```diff
@@ -172,119 +172,79 @@ def find_qemu_binary(config: Config) -> str:
-    ]
+class OvmfConfig(NamedTuple):
+    firmware: Path
+    format: str
+    vars: Path
+    vars_format: str
 
-    for firmware in FIRMWARE_LOCATIONS:
-        if (config.tools() / firmware).exists():
-            return Path("/") / firmware, True
-
-    FIRMWARE_LOCATIONS = [
-        "usr/share/edk2/ovmf/OVMF_CODE.fd",
-        "usr/share/edk2-ovmf/OVMF_CODE.fd",
-        "usr/share/qemu/OVMF_CODE.fd",
-        "usr/share/ovmf/OVMF.fd",
-        "usr/share/OVMF/OVMF_CODE_4M.fd",
-        "usr/share/OVMF/OVMF_CODE.fd",
-    ]
 
-    for firmware in FIRMWARE_LOCATIONS:
-        if (config.tools() / firmware).exists():
-            logging.warn("Couldn't find OVMF firmware blob with secure boot support, "
-                         "falling back to OVMF firmware blobs without secure boot support.")
-            return Path("/") / firmware, False
+def find_ovmf_firmware(config: Config) -> OvmfConfig:
+    desc = flatten(
+        p.glob("*")
+        for p in (
+            Path(os.getenv("XDG_CONFIG_HOME", INVOKING_USER.home() / ".config")) / "qemu/firmware",
+            Path("/etc/qemu/firmware"),
+            Path("/usr/share/qemu/firmware"),
+        )
+    )
 
-    die("Couldn't find OVMF UEFI firmware blob.")
+    arch = config.architecture.to_qemu()
+    machine = config.architecture.default_qemu_machine()
 
+    for p in sorted(desc):
+        j = json.loads(p.read_text())
 
-def find_ovmf_vars(config: Config) -> Path:
-    OVMF_VARS_LOCATIONS = []
+        if "uefi" not in j["interface-types"]:
+            logging.debug(f"{p.name} firmware description does not target UEFI, skipping")
+            continue
 
-    if config.architecture == Architecture.x86_64:
-        OVMF_VARS_LOCATIONS += [
-            "usr/share/ovmf/x64/OVMF_VARS.fd",
-            "usr/share/qemu/ovmf-x86_64-vars.bin",
-            "usr/share/edk2/x64/OVMF_VARS.4m.fd",
-            "usr/share/edk2/x64/OVMF_VARS.fd",
-        ]
-    elif config.architecture == Architecture.x86:
-        OVMF_VARS_LOCATIONS += [
-            "usr/share/edk2/ovmf-ia32/OVMF_VARS.fd",
-            "usr/share/OVMF/OVMF32_VARS_4M.fd",
-            "usr/share/edk2/ia32/OVMF_VARS.4m.fd",
-            "usr/share/edk2/ia32/OVMF_VARS.fd",
-        ]
-    elif config.architecture == Architecture.arm:
-        OVMF_VARS_LOCATIONS += ["usr/share/AAVMF/AAVMF32_VARS.fd"]
-    elif config.architecture == Architecture.arm64:
-        OVMF_VARS_LOCATIONS += ["usr/share/AAVMF/AAVMF_VARS.fd"]
-
-    OVMF_VARS_LOCATIONS += [
-        "usr/share/edk2/ovmf/OVMF_VARS.fd",
-        "usr/share/edk2-ovmf/OVMF_VARS.fd",
-        "usr/share/qemu/OVMF_VARS.fd",
-        "usr/share/ovmf/OVMF_VARS.fd",
-        "usr/share/OVMF/OVMF_VARS_4M.fd",
-        "usr/share/OVMF/OVMF_VARS.fd",
-    ]
+        for target in j["targets"]:
+            if target["architecture"] != arch:
+                continue
+
+            for glob in target["machines"]:
+                if machine in glob:
```

Comment:
Tried that first, doesn't work

Reply:
Ah, yes, the qemu targets I can see in that dir on my machine are `pc-q35-*`, `pc-i440fx-*` and `microvm`, so `fnmatch` can't match this. I think a comment like 
> # "in" is used here because we need to match, e.g. "q35" in "pc-q35-*"

would be helpful when reading this without having to consult the files.